controllers/kafkatopic_controller.go

```diff
@@ -240,108 +258,91 @@ func (r *KafkaTopicReconciler) syncTopicStatus(cluster *banzaicloudv1beta1.Kafka
-	}
-
-	// iterate topic partitions and populate maps with their values
-	leaders := make(map[string]string, 0)
-	replicaCounts := make(map[string]string, 0)
-	isr := make(map[string]string, 0)
-	offlineReplicas := make(map[string]string, 0)
-
-	for _, part := range meta.Partitions {
-		ID := strconv.Itoa(int(part.ID))
-
-		leaders[ID] = fmt.Sprintf("%s/%s", strconv.Itoa(int(part.Leader)), k.ResolveBrokerID(part.Leader))
-		replicaCounts[ID] = strconv.Itoa(len(part.Replicas))
-
-		if len(part.Isr) > 0 {
-			isr[ID] = fmt.Sprintf("%+v", part.Isr)
-		}
-
-		if len(part.OfflineReplicas) > 0 {
-			offlineReplicas[ID] = fmt.Sprintf("%+v", part.OfflineReplicas)
-		}
-
-	}
-	// finally update status
+	// update status
 	updated := topic.DeepCopy()
-	updated.Status = banzaicloudv1alpha1.KafkaTopicStatus{
-		PartitionCount:  int32(len(meta.Partitions)),
-		Leaders:         leaders,
-		ReplicaCounts:   replicaCounts,
-		InSyncReplicas:  isr,
-		OfflineReplicas: offlineReplicas,
-	}
-	if err = r.Client.Status().Update(context.TODO(), updated); err != nil {
+	updated.Status = *status
+	if err = r.Client.Status().Update(ctx, updated); err != nil {
 		syncLogger.Error(err, "Failed to update KafkaTopic status")
 		return true, err
 	}
 	return true, nil
+}
 
+func (r *KafkaTopicReconciler) getKafkaTopicStatus(log logr.Logger, cluster *v1beta1.KafkaCluster, topic *v1alpha1.KafkaTopic) (*v1alpha1.KafkaTopicStatus, error) {
+	// grab a connection to kafka
+	k, close, err := newBrokerConnection(log, r.Client, cluster)
+	if err != nil {
+		return nil, err
+	}
+	defer close()
+
+	// get topic metadata
+	meta, err := k.DescribeTopic(topic.Spec.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	return k.TopicMetaToStatus(meta), nil
 }
 
-func (r *KafkaTopicReconciler) checkFinalizers(reqLogger logr.Logger, broker kafkaclient.KafkaClient, topic *banzaicloudv1alpha1.KafkaTopic) (reconcile.Result, error) {
+func (r *KafkaTopicReconciler) checkFinalizers(ctx context.Context, reqLogger logr.Logger, broker kafkaclient.KafkaClient, topic *v1alpha1.KafkaTopic) (reconcile.Result, error) {
 	reqLogger.Info("Kafka topic is marked for deletion")
 	var err error
 	if util.StringSliceContains(topic.GetFinalizers(), topicFinalizer) {
 		if err = r.finalizeKafkaTopic(reqLogger, broker, topic); err != nil {
 			return requeueWithError(reqLogger, "failed to finalize kafkatopic", err)
 		}
-		if err = r.removeFinalizer(topic); err != nil {
+		if err = r.removeFinalizer(ctx, topic); err != nil {
 			return requeueWithError(reqLogger, "failed to remove finalizer from kafkatopic", err)
 		}
 	}
 	return reconciled()
 }
 
-func (r *KafkaTopicReconciler) removeFinalizer(topic *banzaicloudv1alpha1.KafkaTopic) error {
+func (r *KafkaTopicReconciler) removeFinalizer(ctx context.Context, topic *v1alpha1.KafkaTopic) error {
 	topic.SetFinalizers(util.StringSliceRemove(topic.GetFinalizers(), topicFinalizer))
-	return r.Client.Update(context.TODO(), topic)
+	return r.Client.Update(ctx, topic)
```

Comment:
we should set the typeMeta here as well, like
```
typeMeta := topic.typeMeta
r.Client.Update(ctx,topic)
topic.typeMeta = typeMeta
```
Update method mutates topic and will cause problems since update loses the typeMeta of the config that's used later when setting ownerrefs